### ISSUE TITLE
Quality report

### ISSUE THREAD

**Issue by JohnLukeBentley** [retorquere/zotero-better-bibtex#751]
If I'm not misunderstanding how things should work ...

When using Zotero-better-bibtext (ZBB) to export via Biblatex (Zotero > Tools > Preferences > Export > Default Format: Better BibLaTex) and I'm exporting a `@reference` or `@inreference` type then the ZBB quality report includes misinformation.

For example, for a @reference type I get the following ...

    @reference{sep_2016_stanford,
      author = {{SEP}},
      date = {2016-24},
      title = {The Stanford Encyclopedia of Philosophy},
      note = {biblatexcitekey[sep\_2016\_book];},
      urldate = {2016-12-31},
      editor = {Zalta, Edward N.},
      publisher = {{Metaphysics Research Lab, Stanford University}},
      url = {https://plato.stanford.edu/archives/win2016/},
      edition = {Winter 2016},
      timestamp = {2017-10-19T04:54:57Z}
    }
    % Quality Report for sep_2016_stanford:
    % Missing required field collection

See Report ID: 55DNGP7Q.

(A separate issue is that in the `Extra` field in Zotero I have `biblatexcitekey[sep_2016_book];biblatex[referencetype=reference]` but the string after the semi-colon is not exported into the `note` field - I can't recall if this is by design).

For a @inreference type I get the following ...

    @inreference{wenar_2015_rights,
      author = {Wenar, Leif},
      date = {2015-09-09},
      shorthand = {Wenar in SEP 2015},
      title = {Rights},
      booktitle = {The Stanford Encyclopedia of Philosophy},
      addendum = {First published 2005-12-19.},
      note = {bibtex*: wenar\_2015\_rights},
      urldate = {2016-06-20},
      editor = {Zalta, Edward N.},
      url = {http://plato.stanford.edu/archives/fall2015/entries/rights/},
      edition = {2015 Fall},
      timestamp = {2017-08-13T08:21:57Z}
    }
    % Quality Report for wenar_2015_rights:
    % Missing required field incollection

See Report ID: TDWWV3EP

(In the `Extra` field in Zotero I have 

```
    bibtex[shorthand=Wenar in SEP 2015;addendum=First published 2005-12-19.]
bibtex*: wenar_2015_rights
```

... this also gets chopped up).

What I expect in each case is there to be no quality report as, if I understanding things correctly, there is no `collection` and no `incollection` field.

From the biblatex documentation (biblatex.pdf), page 7 Section 2.1 Entry Types, we have `collection` and `incollection` **Entry Types** but we don't have similarly named **Entry Fields**. See also page 13, Section 2.2 Entry Fields.

Perhaps this is an artefact of a former time when there were such fields,

If we look at page 25, Seciton 2.2.3 Special Fields, there are a number of fields that might play the roles that are currently intimated by the Quality Report for the alleged fields `incollection` and `collection`. Namely, `crossref`, `entryset`, `relatedtype`, `relatedtype`, `relatedstring`, `xref`.

I'm not currently across how those special fields work, so I'm not clear if the best thing here for the Quality report, for "Missing require field [incollection|collection]" is to:

* Completely remove that warning; or
* Substitute an alternative.

Let me know if you want me to do that research.

I'm on Zotero 5.0.23, ZBB 5.0.28.

**Comment by retorquere** [comment]
:robot: bleep bloop; this is your friendly neighborhood build bot announcing [test build 4498](https://github.com/retorquere/zotero-better-bibtex/releases/download/builds/zotero-better-bibtex-5.0.28.4498.issue-751.xpi) ("fixes #751").

**Comment by retorquere** [comment]
:robot: bleep bloop; this is your friendly neighborhood build bot announcing [test build 4499](https://github.com/retorquere/zotero-better-bibtex/releases/download/builds/zotero-better-bibtex-5.0.28.4499.issue-751.xpi) ("scrub old dynamic keys").

**Comment by retorquere** [comment]
I simply messed this up; the biblatex manual says `reference` can be treated as an alias of `collection`, not that it *has* a `collection`. Same goes for `inreference`. I changed the required fields for `inreference` and `reference`. Things should be better now in 4499; that build will also remove the `bibtex*:` stuff from the `extra` field which was an early experiment to get rid of the shadow DB BBT maintains. That latter part is really, really slow so I only enable it whenever I find someone who still has that cruft in the `extra` field, it won't be there in the actual (pre-)releases.

**Comment by retorquere** [comment]
:robot: bleep bloop; this is your friendly neighborhood build bot announcing [test build 4501](https://github.com/retorquere/zotero-better-bibtex/releases/download/builds/zotero-better-bibtex-5.0.28.4501.issue-751.xpi) ("more QR failings").

**Comment by retorquere** [comment]
There were actually a lot more of these problems; 4501 aism to catch them all, but I wouldn't half mind you looking over the config for the QRs for [biblatex](https://github.com/retorquere/zotero-better-bibtex/blob/751/resource/bibtex/biblatex.entry-types.json) and [bibtex](https://github.com/retorquere/zotero-better-bibtex/blob/751/resource/bibtex/bibtex.entry-types.json). The biblatex ones are now corrected to what I found in the biblatex manual; the bibtex ones are just what I had. No `optional` means any field will be accepted without warning.

**Comment by JohnLukeBentley** [comment]
Using 4501 and just on a Biblatex export it looks you have indeed:

* Solved the original bug. There's no `% Missing required field collection` report;
* Removed the `bibtex*` stuff.

What I get now for the prior entries ...

... Report ID: 55DNGP7Q ....

    @reference{sep_2016_stanford,
      author = {{SEP}},
      date = {2016-24},
      title = {The Stanford Encyclopedia of Philosophy},
      note = {biblatexcitekey[sep\_2016\_book];},
      urldate = {2016-12-31},
      editor = {Zalta, Edward N.},
      publisher = {{Metaphysics Research Lab, Stanford University}},
      url = {https://plato.stanford.edu/archives/win2016/},
      edition = {Winter 2016},
      timestamp = {2017-10-19T04:54:57Z}
    }
    % Quality Report for sep_2016_stanford:
    % Unexpected field timestamp
    % Unexpected field author

"Unexpected field author" teaches me that, indeed, the biblatex @reference type doesn't have an author field. So that really sells me on the value of the quality report in general.

Remind me on `timestamp`: that's a field for your use (taking the value from `Modified` in Zotero)? If that's right perhaps the quality string should be more expansive. e.g. "Unexpected field timestamp: not a biblatex field but used by zotero-better-bibtex"

Report ID: TDWWV3EP ...

    @inreference{wenar_2015_rights,
      author = {Wenar, Leif},
      date = {2015-09-09},
      shorthand = {Wenar in SEP 2015},
      title = {Rights},
      booktitle = {The Stanford Encyclopedia of Philosophy},
      addendum = {First published 2005-12-19.},
      urldate = {2016-06-20},
      editor = {Zalta, Edward N.},
      url = {http://plato.stanford.edu/archives/fall2015/entries/rights/},
      edition = {2015 Fall},
      timestamp = {2017-10-26T03:44:43Z}
    }
    % Quality Report for wenar_2015_rights:
    % Unexpected field timestamp
    % Unexpected field shorthand

From biblatex.pdf, page 7 ...

> Generic fields like `abstract` and `annotation` or `label` and `shorthand` are not included in the lists below because they are independent of the entry type. The special fields discussed in § 2.2.3, which are also independent of the entry type, are not included in the lists either.

So you probably want to remove any "expected field" report for all those (4 + 22) fields.

> I wouldn't half mind you looking over the config for the QRs for biblatex and bibtex.

I'll be keen to, after your fix and comments on the above. Thanks for ploughing through the rewrite!

**Comment by retorquere** [comment]
:robot: bleep bloop; this is your friendly neighborhood build bot announcing [test build 4504](https://github.com/retorquere/zotero-better-bibtex/releases/download/builds/zotero-better-bibtex-5.0.28.4504.issue-751.xpi) ("Merge branch 'master' into 751").

**Comment by retorquere** [comment]
> "Unexpected field author" teaches me that, indeed, the biblatex @reference type doesn't have an author field. So that really sells me on the value of the quality report in general.

I implemented QR right after I needed it -- I was docked points on my final thesis because of errors in my references and missed CL by a hair.

> Remind me on `timestamp`: that's a field for your use (taking the value from `Modified` in Zotero)? If that's right perhaps the quality string should be more expansive. e.g. "Unexpected field timestamp: not a biblatex field but used by zotero-better-bibtex"

... which tells you how often I've used it myself. It's going to become really relevant to me soon though.

The timestamp field is an accommodation for JabRef. I'm torn on what to do with it now. Options are to not export it by default (but that would add another option to the already baroque preferences screen), or to add it to the `"*": "allowed"` portion of the (now updated) config (meaning the QR lies a bit compared to the biblatex manual)

>      booktitle = {The Stanford Encyclopedia of Philosophy},

ah, good ole' SEP

>    % Unexpected field shorthand
>
> From biblatex.pdf, page 7 ...

I've added the generic fields at the bottom of the config, under `"*"`. They will apply to all reference types.

> So you probably want to remove any "expected field" report for all those (4 + 22) fields.

You mean "unexpected field", yeah?

> I'll be keen to, after your fix and comments on the above. Thanks for ploughing through the rewrite!

It makes for interesting times, although mostly interesting in the apocryphal Chinese sense. Updates in 4505.

**Comment by retorquere** [comment]
:robot: bleep bloop; this is your friendly neighborhood build bot announcing [test build 4505](https://github.com/retorquere/zotero-better-bibtex/releases/download/builds/zotero-better-bibtex-5.0.28.4505.issue-751.xpi) ("minor cleanups & speedups").

**Comment by JohnLukeBentley** [comment]
## Entry independent fields

On 4505 `shorthand` is not longer reported by the QR. From biblatex.entry-types.json I see your `"*"` entry and that you've included some fields in addition to the one's I suggested ...  that save myself cognitive effort on this sleepy afternoon I'll just assume you've seen further than me on this topic. In short I'm happy to call it that that false positives no longer exist.

## Timestamp as an option

Thanks for the info on timestamp.

> I'm torn on what to do with it now. Options are to not export it by default (but that would add another option to the already baroque preferences screen), ....

I'd be a favour of having an option for this: so that I could turn it off. If it is an option then the issue of whether you have it on or off by default becomes less fraught: a user can just choose their preferred option. So, if it is an option then leaving it on by default might be the thing to do ... to advertise its existence.

## Is QR for biblatex/bibtex only?

Whatever you decide the issue of what to do in the QR (for timestamp) seems separate ....

> or to add it to the "*": "allowed" portion of the (now updated) config (meaning the QR lies a bit compared to the biblatex manual).

If the QR is to be understood to be biblatex (or bibtex when doing a bibtex export) strict then it would be better not to lie. So your choice here seems between
* Have a QR that is biblatex (or bibtex when doing a bibtex export) strict; or
* Have a QR that reports on biblatex (or bibtex when doing a bibtex export) and other considerations (e.g. a field that is good for JabRef). 

## bibtex.entry-types.json and biblatex.entry-types.json

On bibtex.entry-types.json and biblatex.entry-types.json

> The biblatex ones are now corrected to what I found in the biblatex manual; ...

... that looks/sounds right.

> the bibtex ones are just what I had ...

Given that bibtex is old school I'm personally disinterested in it.

So I mean just on a quick glance at  bibtex.entry-types.json and biblatex.entry-types.json, rather setting up a barrage of tests, I'm happy to call that good for now.

**Comment by retorquere** [comment]
> On 4505 `shorthand` is not longer reported by the QR. From `biblatex.entry-types.json` I see your `"*"` entry and that you've included some fields in addition to the one's I suggested ...  that save myself cognitive effort on this sleepy afternoon I'll just assume you've seen further than me on this topic. In short I'm happy to call it that that false positives no longer exist.

I wouldn't overestimate how much sleep I'm getting :roll_eyes:.

> I'd be a favour of having an option for timestamp: so that I could turn it off. If it is an option then the issue of whether you have it on or off by default becomes less fraught: a user can just choose their preferred option. So, if it is an option then leaving it on by default might be the thing to do ... to advertise its existence.

But I'd really rather not add an option per each individual field. We've found this one now, but I can't say there aren't/won't be more of these.

> Is QR for biblatex/bibtex only?

Yes. My intent with the QR is to warn of incomplete/malformed references to avoid getting incorrect references in the final article as it is rendered. The example you mentioned (the `reference` with an `author` field) could indicate you have a reference with the wrong type; maybe you *did* mean to include the author, and you want `article` instead.

Seen that way, it should be safe to just ignore `timestamp`, so

> or to add it to the "*": "allowed" portion of the (now updated) config (meaning the QR lies a bit compared to the biblatex manual).

wouldn't be so bad in that light. It's highly unlikely that having the `timestamp` will mess things up, and since the timestamp will not render anyhow, the `reference`/`article` error possibility doesn't apply.

> If the QR is to be understood to be biblatex (or bibtex when doing a bibtex export) strict then it would be better not to lie. So your choice here seems between
> * Have a QR that is biblatex (or bibtex when doing a bibtex export) strict; or
> * Have a QR that reports on biblatex (or bibtex when doing a bibtex export) and other considerations (e.g. a field that is good for JabRef). 

I'm not too concerned about JabRef. I accomodate it when it's low-impact because enough people want to do some kind of post-processing in it, but I don't personally use it, and my aim with BBT is to enable a workflow that requires no tools except Zotero, BBT, and the latex toolchain.

**Comment by retorquere** [comment]
:robot: bleep bloop; this is your friendly neighborhood build bot announcing [test build 4512](https://github.com/retorquere/zotero-better-bibtex/releases/download/builds/zotero-better-bibtex-5.0.28.4512.issue-751.xpi) ("Merge branch 'master' into 751").

**Comment by JohnLukeBentley** [comment]
> I wouldn't overestimate how much sleep I'm getting 

Well let's muddle through these issues together in the same half asleep state :)

## origdate

`origdate` should probably be added to your "*" list of entry independent fields. I'm asking the biblatex folks why the documentation doesn't mention it in the context of fields that entries support at https://github.com/plk/biblatex/issues/638 

## timestamp

> [Out of order] The timestamp field is an accommodation for JabRef.  .... It's going to become really relevant to me soon though. ... I'm not too concerned about JabRef.  I accomodate it ....  but I don't personally use it ....

Do you mean that `timestamp`:
* Has two separate purposes: for JabRef users; and for some kind of BBT internal purpose (that will become relevant for you, as a dev, soon); or
* It has one purpose: to accommodate JabRef (and you'll, as a user, be forced to temporarily use JabRef soon).

?

> But I'd really rather not add an option per each individual field. We've found this one now, but I can't say there aren't/won't be more of these.

From a usability point of view I don't see a problem with this. I mean I'd agree with the general principle that you don't want to add options unless you need to. But sometimes options provide a way out of a problem. In this case `timestamp`, being an accommodation for something other than biblatex/bibtex, seems exceptional enough. And if there are other such exceptional fields I don't see any usability with adding them to the options UI.

> >  Is QR for biblatex/bibtex only?

> Yes.

Then "lying", omitting a "% Unexpected field 'timestamp'" when a `timestamp` is not biblatex/bibtex compliant, doesn't seem like a good thing. That's building in a conceptual inconsistency into the design and, worse, deliberately hiding that inconsistency. Such hidden inconsistencies, as a general matter, have a tendency to come back to bite you as a dev ... to say nothing of potential confusions for users. I'd suggest that if QR is meant to be for biblatex/bibtex only, it should be strictly so. And I do think it probably a good idea for QR to be for biblatex/bibtex only.

As I say, another option is to have it that QR has exceptions for non biblatex/bibtex purposes, e.g.

> % Unexpected field 'timestamp' (but necessary for JabRef).

Tangentially (and trivially), instead of ...

> % Quality Report for simpson_cool_book_0130: ...

... the following might better advertise the purpose of the reports ....

> % Biblatex Quality Report for simpson_cool_book_0130: ...

.... or when using a Bibtex ....
> % Bibtex Quality Report for simpson_cool_book_0130: ...

Anyway and in summary, I hope my arguments are enough to convince you or push you more clearly in the other direction. 

Even, to think of a third set of choices, making no changes seems plausible. 

Indeed if it were a choice between:

* "Lying": leaving "% Unexpected field 'timestamp'" out of the QR ... and misleading inexperienced users into thinking "timestamp" is biblatex compliant (and having it simply ignored by biblatex); and
* Having "% Unexpected field 'timestamp'" remain ... forcing inexperienced users to learn that `timestamp` is not biblatex complaint (but will be simply ignored by biblatex) ....

... then I think the second choice superior.

**Comment by retorquere** [comment]
> `origdate` should probably be added to your "*" list of entry independent fields. I'm asking the biblatex folks why the documentation doesn't mention it in the context of fields that entries support at https://github.com/plk/biblatex/issues/638 

It would also be possible to just test for both `X` and `origX` in the allowed fields.

> Do you mean that `timestamp`:
> * Has two separate purposes: for JabRef users; and for some kind of BBT internal purpose (that will become relevant for you, as a dev, soon); or
> * It has one purpose: to accommodate JabRef (and you'll, as a user, be forced to temporarily use JabRef soon).

Ah sorry no: I meant that I may be getting back into the academic fray soon so reference quality would become a life concern for me again.

The timestamp has only one purpose (although I hadn't realized it before we had this discussion) which is to accommodate JabRef. But JabRef doesn't strictly require it, and I don't use JabRef at all, I just know others use (or used) JabRef on the files produces by BBT for some kind of post-processing.

> From a usability point of view I don't see a problem with this. I mean I'd agree with the general principle that you don't want to add options unless you need to. But sometimes options provide a way out of a problem. In this case `timestamp`, being an accommodation for something other than biblatex/bibtex, seems exceptional enough. And if there are other such exceptional fields I don't see any usability with adding them to the options UI.

I've changed the `jabrefGroups` pref to `jabrefFormat` pref, and setting/disabling that will now also affect the timestamp being exported at all.

> % Unexpected field 'timestamp' (but necessary for JabRef).

This is going to be in 4522 once I adjust the test cases.

> % Biblatex Quality Report for simpson_cool_book_0130: ...

Also in 4522

BTW I'm scanning through https://github.com/plk/biblatex/issues/638 and I understand some of what is being discussed; I am a fairly unsophisticated latex user, and not familiar with the intricacies of biber et al at all.

**Comment by blip-bloop** [comment]
:robot: this is your friendly neighborhood build bot announcing [test build 4525](https://github.com/retorquere/zotero-better-bibtex/releases/download/builds/zotero-better-bibtex-5.0.28.4525.issue-751.xpi) ("adjust tests for #751").

**Comment by retorquere** [comment]
So make that 4525 instead of 4522.

**Comment by retorquere** [comment]
Do you have any idea what the info in the BCF file means? I can't make anything off it, and I can't find docs.

**Comment by JohnLukeBentley** [comment]
I think I can read the `.bfc` file. But I've sought clarification at https://github.com/plk/biblatex/issues/638#issuecomment-340312148.

I also have all sorts of XML parsing experience ... if that became necessary (although a manual cut and paste, which I think you are already doing, might be the time expedient way to do things in the short term).

Thanks for clarifying the `timestamp` purpose. ...

> The timestamp has only one purpose (although I hadn't realized it before we had this discussion) which is to accommodate JabRef.

Yeah, it's great how talking things through clarifies.

On build 4525 I see:

* When extensions.zotero.translators.better-bibtex.jabrefFormat is set to 0 then the `timestamp` field is not exported (and therefore doesn't trigger a QR about it). Great!
* When extensions.zotero.translators.better-bibtex.jabrefFormat is set to 4 (the default) then the `timestamp` field is exported and the QR flags the `timestamp` field with  "% Unexpected field 'timestamp' (JabRef-specific)". Great!

Is your intention with extensions.zotero.translators.better-bibtex.jabrefFormat to keep that as an advanced config option without a corresponding UI element? 

Is the integer value "4" a bit mask, representing "100"? If so, what's the other two positions for?

Anyway it looks you've headed in the right direction with this. Thanks for the coding!

Good luck, by the way, as you get into your Phd. I'm jealous ... but don't envy the pressure (I prefer to delude myself into thinking there's no deadline when, in fact, there's at least one big one at the very end ... that we all face).

**Comment by retorquere** [comment]
> I also have all sorts of XML parsing experience ... if that became necessary (although a manual cut and paste, which I think you are already doing, might be the time expedient way to do things in the short term).

XML parsing won't be a problem -- it's just that I didn't know what the data meant. I think I have it now.

> Is your intention with extensions.zotero.translators.better-bibtex.jabrefFormat to keep that as an advanced config option without a corresponding UI element? 

I am going to add an option for it.

> Is the integer value "4" a bit mask, representing "100"? If so, what's the other two positions for?

Nope, it's just that the "old" jabref file format had a `groupsformat:3` and the newer version doesn't have any kind of indication and I've just dubbed it `4` as in `post-3`. Jabref doesn't have an internal version numbering for their format and it doesn't correspond cleanly to a major version number.

> Good luck, by the way, as you get into your Phd. I'm jealous ... but don't envy the pressure 

I know it will be bad -- it was pretty bad for the masters. We'll see. I just miss learning terribly.

> (I prefer to delude myself into thinking there's no deadline when, in fact, there's at least one big one at the very end ... that we all face).

Not me baby; I'm going to live forever until the day I die.

**Comment by blip-bloop** [comment]
:robot: this is your friendly neighborhood build bot announcing [test build 4529](https://github.com/retorquere/zotero-better-bibtex/releases/download/builds/zotero-better-bibtex-5.0.28.4529.issue-751.xpi) ("this is going to spiral out of control...").

**Comment by JohnLukeBentley** [comment]
I take it you'll need to release another test build in virtue of catching up with the latest `.bcf` (as pointed to in https://github.com/plk/biblatex/issues/638#issuecomment-340617893).

**Comment by retorquere** [comment]
Oh yeah ignore 4529. I'm going a different tack with the QR; the rules in the BCF are not expressible in the current config file so I'm just going to compile the bcf straight into javascript.

**Comment by blip-bloop** [comment]
:robot: this is your friendly neighborhood build bot announcing [test build 4541](https://github.com/retorquere/zotero-better-bibtex/releases/download/builds/zotero-better-bibtex-5.0.28.4541.issue-751.xpi) ("bcf-based QR").

**Comment by retorquere** [comment]
4541 is based directly on the BCF; the QR is a little different because of it.

**Comment by JohnLukeBentley** [comment]
Thanks!

`origdate` is wrongly flagged. But this looks like a matter of both of us learning more about the structure of `*.bcf`. I've sought clarification at https://github.com/plk/biblatex/issues/638#issuecomment-341266841

**Comment by retorquere** [comment]
My bad. New build underway.

**Comment by blip-bloop** [comment]
:robot: this is your friendly neighborhood build bot announcing [test build 4551](https://github.com/retorquere/zotero-better-bibtex/releases/download/builds/zotero-better-bibtex-5.0.28.4551.issue-751.xpi) ("add generic constraints to allowed").

**Comment by retorquere** [comment]
Should I undo the changes in 4551 given the discussion over at https://github.com/plk/biblatex/issues/638#issuecomment-341266841?

**Comment by blip-bloop** [comment]
:robot: this is your friendly neighborhood build bot announcing [test build 4565](https://github.com/retorquere/zotero-better-bibtex/releases/download/builds/zotero-better-bibtex-5.0.28.4565.issue-751.xpi) ("Merge branch 'master' into 751").

**Comment by blip-bloop** [comment]
:robot: this is your friendly neighborhood build bot announcing [test build 4566](https://github.com/retorquere/zotero-better-bibtex/releases/download/builds/zotero-better-bibtex-5.0.28.4566.issue-751.xpi) ("bcf is... interesting").

**Comment by retorquere** [comment]
alright -- you know the drill.

**Comment by retorquere** [comment]
I've been considering adding an extra line to the QR to warn/explain that when a reference doesn't have a langid, BBT will change titles in sentence case (which is what Zotero recommends doing because citeproc expects it that way) to Title Case (which is what biblatex expects for English references but English references only).

The whole matter of title casing between Zotero and biblatex is a hilarious mess which I won't bother you with right now. The reason why I am considering adding a line is because a lot of people don't understand why the title casing happens at all, and I'm sort of tired of explaining.

The QR will get a preference in the prefs screen BTW; I'm leaning to setting it to default "on".

**Comment by JohnLukeBentley** [comment]
Great work! Probably all of the false positives (like `origdate`) are now absent from the QR.

The new string ....
> % == BibLateX quality report for [citekey] ....

... is more informative (although perhaps "for [citekey]" is redundant).

I hit 4566 with a barrage of examples. The examples are targeting EDTF variations so they are only incidentally, but sufficiently I reckon, testing the Quality Report.

I'm only getting the following fields as flagged (I've turned off `timestamp`):

* `rights`. In an `article` entrytype. See Report ID: QHSNGV9E (barker_2016_turkey)
* `journaltitle` and `type`.  In an `online` entrytype. Report ID: Y4ISV8TP (greenwald_2017_how) ... this is a tweet.

For the `online` entrytype neither `journaltitle` and `type` in supported according to biblatex.pdf (p 10). And the `rights` field seems to be nowhere listed in biblatex.pdf (Indeed searching for the 'rights' string turns up empty). So it looks like your quality report is correctly flagging these fields. That is, these aren't false positives.

However, that that might raise an issue for the biblatex folks; or else I need to think more about those fields.

In summary, the quality report seems to now work as expected on the initial barrage of cannons. You could go ahead and close the issue if the above doens't raise any concerns from your point of view. 

Perhaps we could reopen this thread if obscure future test cases warrant it.

**Comment by retorquere** [comment]
Super.

The way it's set up now, if you get me a new BCF, BBT will adapt. I'm compiling the BCF into Javascript code directly.

**Comment by JohnLukeBentley** [comment]
That's robust and therefore awesome.

**Comment by JohnLukeBentley** [comment]
Suggest to https://github.com/retorquere/zotero-better-bibtex/wiki/Configuration#hidden-preferences something like the following be added ...

> extensions.zotero.translators.better-bibtex.jabrefFormat
> _default: `4`_
>
> Better biblatex will export a `timestamp` field, as in `timestamp = {2017-10-19T04:54:57Z}`. To turn this off set to `0`.

And/or expose this field in the UI.

**Comment by retorquere** [comment]
I don't know which ones I'll add to the UI, but I have code nearly done to auto-generate the docs from the UI code.

**Comment by JohnLukeBentley** [comment]
Ok. :)
